### ISSUE TITLE
Use platform apis for Internationalizion & deprecate Angular Locale getters 

### ISSUE THREAD

**Issue by JeanMeche** [angular/angular#54470]
### Which @angular/* package(s) are the source of the bug?

common

### Is this a regression?

No

### Description

In 2017 (#18284), it was chosen to drop the `Intl` API in favor of custom i18N data extracted from the CLDR (`@angular/common/locales/...`. 

Since then the `Intl` API has matured a lot and is now able to cover the feature provided by the framework. 

Replacing the legacy i18n implementation with the `Intl` API is move to shift the localisation responsability from the framework to the platform (browsers, Node or Cloud workers). 

## Deprecations 
This will lead to the deprecation of the following locale getters : 

* `getCurrencySymbol(code: string, format: 'wide' | 'narrow', locale?: string)`
* `getLocaleCurrencyCode(locale: string)`
* `getLocaleCurrencyName(locale: string)`
* `getLocaleCurrencySymbol(locale: string)`
* `getLocaleDateFormat(locale: string, width: FormatWidth)`
* `getLocaleDateTimeFormat(locale: string, width: FormatWidth)`
* `getLocaleDayNames(locale: string, formStyle: FormStyle, width: TranslationWidth)`
* `getLocaleDayPeriods(locale: string, formStyle: FormStyle, width: TranslationWidth)`
* `getLocaleDirection(locale: string)`
* `getLocaleEraNames(locale: string, width: TranslationWidth)`
* `getLocaleExtraDayPeriodRules(locale: string): (Time | [Time, Time])`
* `getLocaleExtraDayPeriods(locale: string, formStyle: FormStyle, width: TranslationWidth)`
* `getLocaleFirstDayOfWeek(locale: string)`
* `getLocaleId(locale: string)`
* `getLocaleMonthNames(locale: string, formStyle: FormStyle, width: TranslationWidth)`
* `getLocaleNumberFormat(locale: string, type: NumberFormatStyle)`
* `getLocaleNumberSymbol(locale: string, symbol: NumberSymbol)`
* `getLocalePluralCase: (locale: string) => (value: number)`
* `getLocaleTimeFormat(locale: string, width: FormatWidth)`
* `getLocaleWeekEndRange(locale: string)`
* `getNumberOfCurrencyDigits(code: string)`

## Related issues 

* angular/angular#20487
* angular/angular#33803 
* angular/angular#39606
* angular/angular#42315
* angular/angular#48697
* angular/angular#49143
* angular/angular#52177
* angular/angular#54114
* angular/angular#54320

**Comment by kyubisation** [comment]
I think this is the right direction for all the listed reasons. 👍 

Would it still be possible to customize e.g. date formats/logic?
To explain; Switzerland has four national languages ((swiss-)german, (swiss-)french, italian and romansh) and english as a semi-official language. Switzerland however has no official locale specification, so as far as I know the browser locale data and the CLDR data is just a "best effort" for specifying the locale.
We at SBB (national public transport provider) have e.g. a datetime format that does not match the Intl.DateTimeFormat output.
As we have hundreds of Angular apps at SBB, we would be interested in a solution that is easily applied (e.g. a simple hook provided by a library).

Thank you for the consideration.

**Comment by massic80** [comment]
@kyubisation as of now, you could use [Luxon](https://moment.github.io/luxon) and [Angular Luxon](https://www.npmjs.com/package/luxon-angular): you can both configure the [timezone](https://moment.github.io/luxon/#/zones) and [BCP-47 language code](https://moment.github.io/luxon/#/intl?id=how-locales-work) (just like it-CH or rm-CH).

**Comment by kyubisation** [comment]
@massic80 Thank you for the reference, but we would prefer using the native Date object (or in the future the temporal objects).

**Comment by JeanMeche** [comment]
@kyubisation 
Yes we aim to keep our API surface as is so custom formatting will still be supported. 
There will be some minor regressions though as Intl does not support everything (like some standalone formats, defined in the extra locales). 
You'll have a good overview by checking #55283 and more specifically `packages/common/test/i18n/format_date_spec.ts` which illustrates the differences we'll have between the current API and the Intl based one.

**Comment by massic80** [comment]
> @massic80 Thank you for the reference, but we would prefer using the native Date object (or in the future the temporal objects).

You're welcome, I understand. I proposed it cause it's a "wrapper" for Intl, nothing "new" :)

**Comment by daiscog** [comment]
While I don't object to this move to `intl`, I do have concerns that the lack of browser support for some locales will mean features that used to "just work" in Angular no longer will.

For example, Chrome currently does not provide support for the Welsh language, [among others](https://issues.chromium.org/issues?q=%5Bmore_locales%5D), when using `Intl`, but currently Angular does.  A Welsh language date/time picker, for example, will no longer render correctly in Chrome if I replace use of deprecated functions like `getLocaleDayNames` with `Intl`-based code.

I think as an opinionated framework, Angular should document how best to replace uses of all deprecated functions, including how to check for runtime support for the target locale, and how to handle cases where the locale data is not available (e.g., perhaps by dynamically checking during initialization and conditionally loading a polyfill).

**Comment by kyubisation** [comment]
I have had a look at the implementation. 
From my understanding, there is no possibility to "patch" the Angular Intl logic.
This was previously possible, although not officially supported but suggested [here](https://github.com/angular/angular-cli/issues/17752#issuecomment-726204162), by patching the CLDR data.

As an example; The `weekday` format `short` from the Intl.DateTimeFormat renders three letters `Mon`, but company policy specifies two letters `Mo`.

As mentioned, we would love to have a configuration possibility e.g. via DI, which is not possible at the moment, as e.g. the DatePipe directly calls `formatDate`.
Again, thank you for your consideration in this matter.

**Comment by Ketec** [comment]
getWeekInfo does not seem to be supported at all yet.
https://caniuse.com/mdn-javascript_builtins_intl_locale_getweekinfo
firefox has no support, other browsers use non standard implementation.

**Comment by Ketec** [comment]
getLocaleFirstDayOfWeek
Date-fns does not have a way to get the value without hard-coded locale imports. You would literally have to import every locale manually.

It shouldn't be marked as deprecated when you opt out of the new system.

**Comment by JeanMeche** [comment]
@Ketec Angular has the same limitations, It needs to provide all those values via static imports. 

Luxon for example uses a default fallback for the cases where the API isn't supported: 

https://github.com/moment/luxon/blob/cea7b5f7eac87db57e183f48da97210c47af7034/src/impl/locale.js#L513-L521

**Comment by mauriziocescon** [comment]
Hey! 

I wonder: any chance the PR related to this ticket will ever be merged? Just waiting for a major release to deliver it or something else? 

Thanks a lot for the update!

**Comment by apiweb** [comment]
> Hey!
> 
> I wonder: any chance the PR related to this ticket will ever be merged? Just waiting for a major release to deliver it or something else?
> 
> Thanks a lot for the update!

I also expect this as well. These changes are great and will significantly improve working with dates. 
I also hope that in the future, the Intl platform will support the new [JavaScript Temporal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Temporal) object.

**Comment by Sefibrah** [comment]
Just reminding you that this is still needed...

**Comment by skarps0711** [comment]
Now It's most wanted

**Comment by Xriuk** [comment]
Seeing that at least some of the API (if not all) were already marked as deprecated, I'm here to add that there is no official Intl support for the methods `getLocaleTimeFormat`, `getLocaleDateFormat` and `getLocaleDateTimeFormat` which would return the format string to use for formatting (instead of doing the actual formatting).

**Comment by JeanMeche** [comment]
@Xriuk You should probably use the Intl output directly isn't of really on an intermediate format.

**Comment by PowerKiKi** [comment]
@JeanMeche could you please comment on the status of this issue ? Is it still something Angular would like to solve in core ? or should we look at alternatives solutions from the community, outside of Angular repositories ?

**Comment by JeanMeche** [comment]
None of the famous JS date libraries fully fully on Intl. You'd likely want to implement your own thing wrapper around Intl (and possibly pipes). We don't have a timeframe for landing this feature for now.
